### PR TITLE
HARP-8870 Fix cull frustum setup when terrain data source is used.

### PR DESCRIPTION
**Pull request by MPursche**
Signed-off-by: Marcel Pursche <marcel.pursche@here.com>

Thank you for contributing to harp.gl!

Before requesting a pull request, please remember to check the following documents:
* [contribution guidelines](https://github.com/heremaps/harp.gl/blob/master/CONTRIBUTING.md)
* [coding style](https://github.com/heremaps/harp.gl/blob/master/CODINGSTYLE.md)

If you are adding new functionality we would highly appreciate if you can describe what is the capability you are adding and even better if you can add some examples. Please also remember to add tests for it.

# CI Check

Our bots will check whether your PR can be directly integrated into the mainline. We have some internal integration tests running on the background, our bots will inform you of the next steps and someone from our team will take a look and help if needed!

And please do not forget to sign-off your commit! You can read more about DCO [here](https://julien.ponge.org/blog/developer-certificate-of-origin-versus-contributor-license-agreements/). But, in short, you just need to use `git commit -s` or append `--signoff` when you are committing to the repo.

Happy contributing!

### CODE REVIEW

File: @here/harp-mapview/lib/MapView.ts [hunk 1]
```diff
@@ -2418,11 +2418,12 @@ export class MapView extends THREE.EventDispatcher {
...
-        // Try to remove incase this method was already called, will do nothing if it doesn't exist.
-        this.removeDataSource(elevationSource);
-        this.addDataSource(elevationSource);
+        // Add as datasource if it was not added before
+        const isPresent = this.m_tileDataSources.some((ds) => ds === elevationSource);
+        if (!isPresent) {
+            this.addDataSource(elevationSource);
+        }
         this.m_elevationRangeSource = elevationRangeSource;
-        this.m_elevationRangeSource.connect();
```
Comment (MPursche) [inline @here/harp-mapview/lib/MapView.ts]: This can be removed, because `dataSource.connect()` was already called by `addDataSource()`. Calling it twice creates some errors in the console.

File: @here/harp-mapview/lib/MapView.ts [hunk 2]
```diff
@@ -2418,11 +2418,12 @@ export class MapView extends THREE.EventDispatcher {
         elevationRangeSource: ElevationRangeSource,
         elevationProvider: ElevationProvider
     ) {
-        // Try to remove incase this method was already called, will do nothing if it doesn't exist.
-        this.removeDataSource(elevationSource);
-        this.addDataSource(elevationSource);
+        // Add as datasource if it was not added before
+        const isPresent = this.m_tileDataSources.some(ds => ds === elevationSource);
```
Comment (nzjony) [inline @here/harp-mapview/lib/MapView.ts]: you are slightly modifying the behaviour of the API, because now the user can no longer swap between two elevation sources, once they call this method, the MapView can only use that specific ElevationSource. I don't know if there is a usecase to swap elevation sources, but because it is currently possible in the API, I think it should at least be documented that once an elevatoin source is set, it can't be changed.

I think this is

File: @here/harp-mapview/lib/VisibleTileSet.ts [hunk 3]
```diff
@@ -1190,22 +1190,21 @@ export class VisibleTileSet {
...
 
-        // If elevation is to be taken into account create extended frustum:
+        // If elevation is to be taken into account extend view frustum:
         // (near ~0, far: maxVisibilityRange) that allows to consider tiles that
         // are far below ground plane and high enough to intersect the frustum.
         if (elevationRangeSource !== undefined) {
-            const fp = MapViewUtils.getCameraFrustumPlanes(this.m_frustumIntersection.camera);
-            fp.near = this.m_viewRange.minimum;
-            fp.far = this.m_viewRange.maximum;
-            this.m_projectionMatrixOverride.makePerspective(
```
Comment (nzjony) [inline @here/harp-mapview/lib/VisibleTileSet.ts]: why did you remove makePerspective? I don't see why you need to use a PerspectiveCamera, when makePerspective should be all you need (it seems that you just need to fix the near / far planes)?

File: @here/harp-mapview/lib/MapView.ts (hunk 2 above)
Comment (MPursche) [reply @here/harp-mapview/lib/MapView.ts]: Ok, but that wouldn't have worked before either. The code would have removed the `elevationSource` that was given as a parameter and than add it again. For switching the source we may want to remove the previous `mapView.m_elevationRangeSource`.

File: @here/harp-mapview/lib/VisibleTileSet.ts (hunk 3 above)
Comment (MPursche) [reply @here/harp-mapview/lib/VisibleTileSet.ts]: Newest version of THREE.js uses `top, bottom, left, right, near, far` as parameters for `Matrix4.makePerspective()`. Calculating these parameters from `fov, aspect, near, far` is encapsuled in the `PerspectiveCamera`. I didn't want to reimplement the code to calculate these parameters in harp.gl as THREE.js may change the way it creates the `projectionMatrix` from the camera parameters or changes the API again.

File: @here/harp-mapview/lib/MapView.ts (hunk 2 above)
Comment (nzjony) [reply @here/harp-mapview/lib/MapView.ts]: Oh you are right, I misread, resolving.

File: @here/harp-mapview/lib/VisibleTileSet.ts (hunk 3 above)
Comment (nzjony) [reply @here/harp-mapview/lib/VisibleTileSet.ts]: I understand, I made the same mistake as the author of this code, thinking that you can change the near / far without adjusting the sides. My bad.
